### PR TITLE
Add new input method: fcitx-libpinyin

### PR DESCRIPTION
**Pull request by yshui**
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [ ] NixOS
  - [ ] OS X
  - [x] Linux
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by ericsagnes** [comment]
Nice, a new fcitx engine incoming!

There is a big issue at the moment though, `libpinyn` build is accessing the network to dowload a file (`http://downloads.sourceforge.net/libpinyin/models/model12.text.tar.gz`) that will make the build fail on Hydra or when `nix.useSandbox` option is set.
This is must be fixed, else `libpinyin` will never have a binary substitute.

The usual way to fix this kind of problem is to fetch the file in the store, copy it in the build source, and make the build not using the network anymore (flags or patch).

A few nitpicks:
- The commit name for the libpinyin update should be `libpinyin: 1.3.0 -> 1.6.0`.
- In `fcitx-libpinyin/default.nix`, `store_path` could have a more explicit name, eg `model_text`.
- If you can write docbook, please add `libpinyin` to the fcitx engines list in the input methods [documentation](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/i18n/input-method/default.xml#L82).

**Comment by yshui** [comment]
fcitx-linpinyin is actually also downloading a file. Which I fixed using the method you mentioned. Will do that to libpinyin too.

File: pkgs/development/libraries/libpinyin/default.nix [hunk 1]
```diff
@@ -10,10 +10,17 @@ stdenv.mkDerivation {
     platforms   = platforms.linux;
   };
 
-  buildInputs = [ glib db pkgconfig ];
+  preConfigure = ''
+    aclocal && libtoolize --force && autoheader && automake -a && autoconf
```
Comment (fpletz) [inline pkgs/development/libraries/libpinyin/default.nix]: You could use `autoreconfHook` instead of manually calling autotools.

**Comment by ericsagnes** [comment]
@yokto I made a PR to update [ibus-libpinyin](https://github.com/NixOS/nixpkgs/pull/19112) that also updated libpinyin. Could you rebase the PR and fix the generated conflicts? sorry for the trouble.

Also, in case you don't feel like being the maintainer for fcitx-libpinyin, feel free to put me as the maintainer.

**Comment by yshui** [comment]
@ericsagnes You just did my work for me, thanks :)

**Comment by ericsagnes** [comment]
@yshui I tried it but getting a segfault as soon as I type a character:

```
Oct 02 13:53:03 nixos xsession[921]: FCITX 4.2.9.1 -- Get Signal No.: 11
Oct 02 13:53:03 nixos xsession[921]: Date: try "date -d @1475383983" if you 
are using GNU date ***
Oct 02 13:53:03 nixos xsession[921]: ProcessID: 965
Oct 02 13:53:03 nixos xsession[921]: 
/nix/store/p9jfbh0acyrrfya22i6ld5pgrp2lhhw2-fcitx-with-plugins-4.2.9.1/bin/fcitx[0x40173f]
Oct 02 13:53:03 nixos xsession[921]: 
/nix/store/6fix3zqpnahyml8zp2sxi2rwan55rgb8-glibc-2.24/lib64/libc.so.6(+0x33240)[0x7f83637d9240]
Oct 02 13:53:03 nixos xsession[921]: 
/nix/store/5v7w3fdaz4whpfz63fwr2bsy09b5spqm-libpinyin-1.6.0/lib/libpinyin.so.9(pinyin_parse_more_full_pinyins+0x62)[0x7f8351b55702]
Oct 02 13:53:03 nixos xsession[921]: 
/nix/store/p9jfbh0acyrrfya22i6ld5pgrp2lhhw2-fcitx-with-plugins-4.2.9.1//lib/fcitx/fcitx-libpinyin.so(+0x5201)[0x7f8351de1201]
Oct 02 13:53:03 nixos xsession[921]: 
/nix/store/vcbkd5wh75825pvjbiclgbs9nla059qm-fcitx-4.2.9.1/lib/libfcitx-core.so.0(FcitxInstanceProcessKey+0x719)[0x7f83643a38f9]
Oct 02 13:53:03 nixos xsession[921]: 
/nix/store/p9jfbh0acyrrfya22i6ld5pgrp2lhhw2-fcitx-with-plugins-4.2.9.1//lib/fcitx/fcitx-xim.so(+0x4d77)[0x7f8350b92d77]
Oct 02 13:53:03 nixos xsession[921]: 
/nix/store/p9jfbh0acyrrfya22i6ld5pgrp2lhhw2-fcitx-with-plugins-4.2.9.1//lib/fcitx/fcitx-xim.so(+0x3e8f)[0x7f8350b91e8f]
Oct 02 13:53:03 nixos xsession[921]: 
/nix/store/p9jfbh0acyrrfya22i6ld5pgrp2lhhw2-fcitx-with-plugins-4.2.9.1//lib/fcitx/fcitx-xim.so(+0xc65c)[0x7f8350b9a65c]
Oct 02 13:53:03 nixos xsession[921]: 
/nix/store/p9jfbh0acyrrfya22i6ld5pgrp2lhhw2-fcitx-with-plugins-4.2.9.1//lib/fcitx/fcitx-xim.so(+0xcdb0)[0x7f8350b9adb0]
Oct 02 13:53:03 nixos xsession[921]: 
/nix/store/p9jfbh0acyrrfya22i6ld5pgrp2lhhw2-fcitx-with-plugins-4.2.9.1//lib/fcitx/fcitx-xim.so(+0xf90b)[0x7f8350b9d90b]
Oct 02 13:53:03 nixos xsession[921]: 
/nix/store/p9jfbh0acyrrfya22i6ld5pgrp2lhhw2-fcitx-with-plugins-4.2.9.1//lib/fcitx/fcitx-x11.so(+0x2d76)[0x7f835b2b0d76]
Oct 02 13:53:03 nixos xsession[921]: 
/nix/store/p9jfbh0acyrrfya22i6ld5pgrp2lhhw2-fcitx-with-plugins-4.2.9.1//lib/fcitx/fcitx-x11.so(+0x3033)[0x7f835b2b1033]
Oct 02 13:53:03 nixos xsession[921]: 
/nix/store/vcbkd5wh75825pvjbiclgbs9nla059qm-fcitx-4.2.9.1/lib/libfcitx-core.so.0(+0x9886)[0x7f8364398886]
Oct 02 13:53:03 nixos xsession[921]: 
/nix/store/vcbkd5wh75825pvjbiclgbs9nla059qm-fcitx-4.2.9.1/lib/libfcitx-core.so.0(FcitxInstanceRun+0x3f)[0x7f83643990bf]
Oct 02 13:53:03 nixos xsession[921]: 
/nix/store/p9jfbh0acyrrfya22i6ld5pgrp2lhhw2-fcitx-with-plugins-4.2.9.1/bin/fcitx[0x40110c]
Oct 02 13:53:03 nixos xsession[921]: 
/nix/store/6fix3zqpnahyml8zp2sxi2rwan55rgb8-glibc-2.24/lib64/libc.so.6(__libc_start_main+0xf0)[0x7f83637c62e0]
Oct 02 13:53:03 nixos xsession[921]: 
/nix/store/p9jfbh0acyrrfya22i6ld5pgrp2lhhw2-fcitx-with-plugins-4.2.9.1/bin/fcitx(_start+0x2a)[0x40118a]
```

Does this happen on your setup too?

**Comment by yshui** [comment]
@ericsagnes Works fine here... 

ping @epico @csslayer

**Comment by yshui** [comment]
@ericsagnes Works fine here... 

ping @epico @wengxt

**Comment by ericsagnes** [comment]
I tested in a VM on a non-stable nixpkgs version, so if it works fine on your side, I guess that segfault is coming from my setup.

All good then.

**Pull request merged**